### DISCUSSION TITLE
Define a set of transformation for a volumes with variable number of slices

### DISCUSSION THREAD

**Discussion by emi-dm** [Project-MONAI/MONAI#7629] [Q&A] [answered]
How can I define a set of transformations for a 3D UNet model given a set of volumes with a variable number of slices? I.e. I have 512x512xA slices and A is variable (e.g. 42, 120, 56).

I tried the transformation with RandCropByPosNegLabel with the parameter allow_smaller=True, but it doesn't work correctly.

Is there already an example in the form of a notebook or file?

Thanks in advanced!

**Comment by KumoLiu** [comment]
Hi @emi-dm, for images with variable slice numbers, you will still apply the set of transformations to each volume, regardless of the slice count. 
`allow_smaller=True` in `RandCropByPosNegLabel` is to allow for when the ROI size is larger than the input image size. Volumes with less slices will be 'smaller' along z-dimension in the voxel space and will be allowed to go for further processing.

Could you please share more information about your use case or perhaps share a small piece of code that can reproduce the issue?

Hope it helps, thanks.

**Reply by emi-dm** [reply]
Hi @KumoLiu !! Thanks for your quick reply :) Here you can reproduce the issue: [https://colab.research.google.com/drive/1J1lAAAnwwSNgIUfpm5kF-GQX6U3n6R31?usp=sharing](https://colab.research.google.com/drive/1J1lAAAnwwSNgIUfpm5kF-GQX6U3n6R31?usp=sharing 
)
I hope that you can help me!

**Reply by emi-dm** [reply]
I don't quite know how to process such a series of volumes with a variable number of slices (e.g. 512x512xN, where N is a variable number of slices). I could use a pad but I don't know if that would alter the learning of the model, could you give me some idea or advice on how I can deal with this case? 

Thanks in advanced!

**Reply by KumoLiu** [reply]
I take a look at the code, I believe the issue isn't associated with the transforms. It's crucial to ensure that the entered size is a multiple of the times the image is downsampled. For instance, if you downsample 4 times, the input should be a multiple of 4.
Thanks.

**Reply by emi-dm** [reply]
Hi, I solve the issue addind a SpatialPadd transform: 

```
import monai

LOWER_BOUND_WINDOW = int(-650 - (1500 // 2)) # TODO: Move this to constants.py
UPPER_BOUND_WINDOW = int(-650 + (1500 // 2))

hrct_transform_3d = monai.transforms.Compose([
    monai.transforms.LoadImaged(keys=["img","seg"], image_only=True, ensure_channel_first=True),
    monai.transforms.RandCropByPosNegLabeld(keys=["img","seg"], label_key="seg", spatial_size=(96,96,96), num_samples=4, allow_smaller=True),
    monai.transforms.SpatialPadd(keys=["img","seg"], spatial_size=(96,96,96), method="end"),
    monai.transforms.ScaleIntensityRanged(keys=["img"], a_min=LOWER_BOUND_WINDOW, a_max=UPPER_BOUND_WINDOW, b_min=0., b_max=1.),
    monai.transforms.ToTensord(keys=["img", "seg"]),
])


cbct_transform_3d = monai.transforms.Compose([
    monai.transforms.LoadImaged(keys=["img","seg"], image_only=True, ensure_channel_first=True),
    monai.transforms.Resized(keys=["img", "seg"], spatial_size=(512,512,-1), mode=("trilinear", "nearest")),
    monai.transforms.RandCropByPosNegLabeld(keys=["img", "seg"], label_key="seg", spatial_size=(96, 96, 96),
                                            num_samples=4, allow_smaller=True),
    monai.transforms.SpatialPadd(keys=["img","seg"], spatial_size=(96,96,96), method="end"),
    monai.transforms.ScaleIntensityd(keys=["img"], minv=0., maxv=255.),
    monai.transforms.ToTensord(keys=["img", "seg"]),
])
```
I have used the "final" method because my aim is to fill images with a depth of less than 96, which are problematic. This is because the crop transformation with the allow_smaller=True parameter does not adapt the depth of images with a shallower depth to 96.  

I have modified the notebook above so you can see it, can you tell me why this might be (I imagine that when you feed the model with the first input, you set that input shape for it)?

**Reply by KumoLiu** [reply, marked as answer]
When the network is downsampled 4 times, the input should ideally be a multiple of 2 to the power of 4, which is 16. So the main point here is that you need ensure all the input be a multiple of 16, 96 does, 32 should also be fine. Thanks.

**Reply by emi-dm** [reply]
Thank you so much @KumoLiu !! You helped me to understand the problem :) I hope that this discussion can help someone
